GillianCore/engine/FOLogic/Reduction.ml

```diff
@@ -2412,638 +2636,84 @@ and substitute_for_list_length (pfs : PFS.t) (le : Expr.t) : Expr.t =
-               fe
-                 (Expr.Infix.( - )
-                    (Expr.list_length left_list)
-                    (Expr.list_length right_list))
-             with
-             | Expr.Lit (Int k) when not (Z.equal k Z.zero) -> true
-             | _ -> false ->
-          (* If we have two lists but can reduce the equality of their lengths to false,
-             then we know the lists cannot be equal*)
-          False
-      | Eq (NOp (LstCat, les), LVar x)
-        when List.mem (Expr.LVar x) les
-             && List.exists
-                  (fun e ->
-                    match e with
-                    | Expr.EList (_ :: _) | Lit (LList (_ :: _)) -> true
-                    | _ -> false)
-                  les -> False
-      | Eq (UnOp (NumToInt, le), re) -> Eq (le, UnOp (IntToNum, re))
-      | Eq (le, UnOp (NumToInt, re)) -> Eq (UnOp (IntToNum, le), re)
-      | And (a1, a2) -> (
-          let fa1 = f a1 in
-          let fa2 = f a2 in
-          match (fa1, fa2) with
-          | False, _ | _, False -> False
-          | True, a | a, True -> a
-          | _, _ -> And (fa1, fa2))
-      | Or (a1, a2) -> (
-          let fa1 = f a1 in
-          let fa2 = f a2 in
-          match (fa1, fa2) with
-          | False, a | a, False -> a
-          | True, _ | _, True -> True
-          | _, _ ->
-              if PFS.mem pfs fa1 || PFS.mem pfs fa2 then True
-              else if PFS.mem pfs (Not fa1) then fa2
-              else if PFS.mem pfs (Not fa2) then fa1
-              else Or (fa1, fa2))
-      (* JOSE: why the recursive call? *)
-      | Not a -> (
-          let fa = f a in
-          match a with
-          | True -> False
-          | False -> True
-          | Not a -> a
-          | Or (a1, a2) -> And (Not a1, Not a2)
-          | And (a1, a2) -> Or (Not a1, Not a2)
-          | FLess (e1, e2) -> FLessEq (e2, e1)
-          | FLessEq (e1, e2) -> FLess (e2, e1)
-          | ILess (e1, e2) -> ILessEq (e2, e1)
-          | ILessEq (e1, e2) -> ILess (e2, e1)
-          | _ -> Not fa)
-      | Eq (e1, e2) -> (
-          let re1 = fe e1 in
-          let re2 = fe e2 in
-          (* Warning - NaNs, infinities, this and that, this is not good enough *)
-          let eq = re1 = re2 in
-          if eq then True
-          else
-            let t1, s1 = Typing.type_lexpr gamma re1 in
-            let t2, s2 = Typing.type_lexpr gamma re2 in
-            if
-              s1 && s2
-              &&
-              match (t1, t2) with
-              | Some t1, Some t2 -> t1 <> t2
-              | _, _ -> false
-            then False
-            else
-              let ite a b : Formula.t = if a = b then True else False in
-              let default re1 re2 : Formula.t = Eq (re1, re2) in
-              match (re1, re2) with
-              (* DEDICATED RPFS REDUCTIONS *)
-              | NOp (LstCat, _), LVar y when rpfs && prefix_catch pfs re1 y ->
-                  Eq (UnOp (LstLen, re1), UnOp (LstLen, re2))
```

Comment:
Removed this, as it incorrectly converted list equalities to list length equalities -- maybe because I removed the `rpfs` param? Either way it didn't seem to make that big of a difference 🤔